connd/qconnectionagent.cpp

```diff
@@ -603,46 +630,55 @@ void QConnectionAgent::openConnectionDialog(const QString &type)
 
 void QConnectionAgent::startTethering(const QString &type)
 {
-    if (!netman | (type != "wifi")) { //we only support wifi for now
-        Q_EMIT tetheringFinished(false);
+    if (!netman | (type != "wifi" && type !="bluetooth")) { // support wifi and bt
         return;
     }
-
+    qCDebug(connAgent) << "startTethering" << type;
     NetworkTechnology *tetherTech = netman->getTechnology(type);
     if (!tetherTech) {
-        Q_EMIT tetheringFinished(false);
-        return;
-    }
-
-    QVector <NetworkService *> services = netman->getServices("cellular");
-    if (services.isEmpty()) {
-        Q_EMIT tetheringFinished(false);
-        return;
-    }
-
-    NetworkService *cellService = services.at(0);
-    if (!cellService || netman->offlineMode()) {
-        Q_EMIT tetheringFinished(false);
+        if (type == "wifi") {
+            Q_EMIT wifiTetheringFinished(false);
+        } else {
+            Q_EMIT btTetheringFinished(false);
+        }
         return;
     }
 
+    bool techPowered = tetherTech->powered();
     QSettings confFile;
     confFile.beginGroup("Connectionagent");
-    bool cellConnected = cellService->connected();
-    bool cellAutoconnect = cellService->autoConnect();
-
-    // save cellular connection state
-    confFile.setValue("tetheringCellularConnected",cellConnected);
-    confFile.setValue("tetheringCellularAutoconnect",cellAutoconnect);
-
-    bool techPowered = tetherTech->powered();
-
-    // save wifi powered state
-    confFile.setValue("tetheringTechPowered",techPowered);
-    confFile.setValue("tetheringType",type);
 
-    delayedTethering = true;
-    tetheringWifiTech = tetherTech;
+    if (type == "wifi") { // Only force cellular on for wifi. Bt can use either when available.
+        QVector <NetworkService *> services = netman->getServices("cellular");
+        if (services.isEmpty()) {
+            Q_EMIT wifiTetheringFinished(false);
+            return;
+        }
+        NetworkService *cellService = services.at(0);
+        if (!cellService || netman->offlineMode()) {
+            Q_EMIT wifiTetheringFinished(false);
+            return;
+        }
+        bool cellConnected = cellService->connected();
+        bool cellAutoconnect = cellService->autoConnect();
+
+        // save cellular connection state
+        confFile.setValue("tetheringCellularConnected",cellConnected);
+        confFile.setValue("tetheringCellularAutoconnect",cellAutoconnect);
+    
+        // save wifi powered state
+        confFile.setValue("tetheringTechPowered",techPowered);
+
+        tetheringWifiTech = tetherTech;
+        delayedTethering = true;
+
+    } else if (type == "bluetooth") {
+        tetheringBtTech = tetherTech;
+        btTetheringEnabled = true;
+        if (!confFile.value("tetheringBtEnabled", false).toBool()) {
+            confFile.setValue("tetheringBtEnabled", true);
+        }
+    }
 
     if (!techPowered) {
```

Comment:
The idea is that if someone turns on Bluetooth tethering while it's switched off, then they're implying that they want it turned on. But it doesn't enable any internet connections if they're both off, because it's not clear which one is intended to be used.